app/models/manageiq/providers/vmware/inventory/parser/cloud_manager.rb

```diff
@@ -0,0 +1,108 @@
+class ManageIQ::Providers::Vmware::Inventory::Parser::CloudManager < ManageIQ::Providers::Vmware::Inventory::Parser
+  def parse
+    vdcs
+    vapps
+    vms
+    vapp_templates
+    images
+  end
+
+  private
+
+  def vdcs
+    collector.vdcs.each do |vdc|
+      persister.availability_zones.find_or_build(vdc.id).assign_attributes(
+        :name => vdc.name
+      )
+    end
+  end
+
+  def vapps
+    collector.vapps.each do |vapp|
+      persister.orchestration_stacks.find_or_build(vapp.id).assign_attributes(
+        :name        => vapp.name,
+        :description => vapp.name,
+        :status      => vapp.human_status,
+      )
+    end
+  end
+
+  def vms
+    collector.vms.each do |vm|
+      parsed_vm = persister.vms.find_or_build(vm[:vm].id).assign_attributes(
+        :uid_ems             => vm[:vm].id,
+        :name                => vm[:vm].name,
+        :hostname            => vm[:hostname],
+        :vendor              => 'vmware',
+        :raw_power_state     => vm[:vm].status,
+        :orchestration_stack => persister.orchestration_stacks.lazy_find(vm[:vm].vapp_id),
+        :snapshots           => []
+      )
+
+      if (resp = vm[:snapshot]) && (snapshot = resp.fetch_path(:body, :Snapshot))
+        parsed_vm.snapshots = [
+          persister.snapshots.find_or_build(parsed_vm.ems_ref).assign_attributes(
```

Comment:
right, this is basically translated to
```.find_or_build_by(:vm_or_template => parsed_vm)```

Reply:
Thanks @Ladas for explanation, I was struggling with this for quite a while and yielded eventually with "well as long as it now works I technically don't need to understand it" 😃 

@Ladas in case you're interested in what the original problem was: so we crashed hard when inventoring brand new snapshots (less than 15.minutes old). Turns out `Snapshot` has [`after_create_callback`](https://github.com/manageiq/manageiq/blob/master/app/models/snapshot.rb#L14-L16) where it already references parent VM. And this is why we crashed: we didn't have VM yet.

Reply:
right, the ```parent_inventory_collections``` will force the order, s othat should fix it (it's also needed for automatic building of targeted queries for targeted refresh)

btw. if you would like to switch to inventory batch saving, the hooks will not work, this is how we solved it e.g. for OpenShift provider https://github.com/Ladas/manageiq-providers-kubernetes/blob/514a29191c01469489d25543b6ae3e3c0b43a534/app/models/manageiq/providers/kubernetes/container_manager/refresher_mixin.rb#L77

(taking all new created records and batching the event raising jobs is also way more effective than the by 1 record hook)

Reply:
https://github.com/ManageIQ/manageiq-providers-vmware/pull/217/files#r183678300 looks :+1: , we can probably drop the ```:ems_ref     => uid,``` from the assign attributes, since we already set it by find_or_build_by

Reply:
The man is right, I've updated the snippet. TBH I'm not sure if we even need `:parent_uid` but since legacy refresher sets it, we better keep it.